external-crates/move/crates/move-package-alt/src/schema/manifest.rs

```diff
@@ -116,6 +129,127 @@ struct RField {
     r: BTreeMap<String, toml::Value>,
 }
 
+#[derive(Error, Debug)]
+#[error("{kind}")]
+pub struct ManifestError {
+    pub kind: Box<ManifestErrorKind>,
+    location: ErrorLocation,
+}
+
+#[derive(Debug)]
+enum ErrorLocation {
+    WholeFile(PathBuf),
+    AtLoc(Location),
+}
+
+#[derive(Error, Debug)]
+pub enum ManifestErrorKind {
+    #[error("package name cannot be empty")]
+    EmptyPackageName,
+    #[error("unsupported edition '{edition}', expected one of '{valid}'")]
+    InvalidEdition { edition: String, valid: String },
+    #[error("externally resolved dependencies must have exactly one resolver field")]
+    BadExternalDependency,
+    #[error(
+        "dep-replacements.mainnet is invalid because mainnet is not in the [environments] table"
+    )]
+    MissingEnvironment { env: EnvironmentName },
+    #[error(
+        // TODO: add a suggested environment (needs to be part of the flavor)
+        "you must define at least one environment in the [environments] section of `Move.toml`."
+    )]
+    NoEnvironments,
+    #[error("{}", .0.message())]
+    ParseError(#[from] toml_edit::de::Error),
+    #[error(transparent)]
+    IoError(#[from] std::io::Error),
+}
+
+impl ParsedManifest {
```

Comment:
This part I like less. The `schema` package is just about the on-disk serialization for (non-legacy) files. I think it would be better to have the validation live in `package`.

The problem is that if we're doing validation, we're establishing some invariants, and then we need to make the fields private to ensure that everything goes through the constructor, and we also end up repeating some of the validation between the new and legacy packages.

If it ends up making `package_impl` too heavy, then maybe we should keep `manifest` (at least the file, if not the type).